### PR TITLE
Split commit to address Maltonns's PR. Add test for SSVQE

### LINKED ISSUE #15: add test
We want to add a test for ssvqecasci

```
def test_ssvqecasci_h2o_2e_2o():
    mc = vqemcscf.VQECASCI(
        mf, 4, 4, optimizer=LBFGS(), ansatz=Ansatz.GateFabric, layers=2,
    )
    mc.fcisolver.nroots = 5
    mc.kernel()

    # refmc = mcscf.CASCI(mf, 4, 4)
    # refmc.fcisolver.nroots = 5
    # refmc.kernel()
    # print(refmc.e_tot)

    ref_energies=[-74.97211297714749, -74.56710399202353, -74.52886408322314, -74.46572136143928, -74.39460452887903]

    assert all(utils.almost_equal(a,b) for a,b in zip(mc.fcisolver.energies, ref_energies))
```

### PR DESCRIPTION
**Pull request by LucaErh**
This is actually to take the bloated PR https://github.com/wmizukami/chemqulacs/pull/12 and make it split commits

This PR adds the ssvqe test

### CODE REVIEW

**Comment by yyoshiyy** [comment]
Please change the title of the pull request, as it may not clear to others what they are about. Specifically, write “this is part of a split commit” and “what the commits are about”.

**Comment by LucaErh** [comment]
The test was indeed not set up correctly. It only produced the ground state energy.
Running the fixed code: 
```
from pyscf import gto, mcscf, scf
from quri_parts.algo.optimizer import LBFGS

from chemqulacs.util import utils
from chemqulacs.vqe import vqemcscf
from chemqulacs.vqe.vqeci import Ansatz

geom_water = utils.get_geometry_from_pubchem("water")
mol = gto.M(atom=geom_water, basis="sto-3g")
mf = scf.RHF(mol)
mf.run()

mc = vqemcscf.VQECASCI(
    mf,
    2,
    2,
    optimizer=LBFGS(),
    ansatz=Ansatz.GateFabric,
    layers=2,
    excitation_number=3
)
mc.kernel()
print("-----Produced energies------------")
print(mc.fcisolver.energies)
```
Produces the output: 
```
converged SCF energy = -74.9644475827695
----VQE-----
Optimizer converged
CASCI E = -74.9656951104499  E(CI) = -1.66322266488406  S^2 = 0.0000000
-----Produced energies------------
[-74.96569511044994, -74.56710399202518, -74.52886408322478, -74.52886408322476]
```

**Comment by LucaErh** [comment]
These are the same energies for which are tested in the ssvqe test.

File: tests/test_vqe_vqemcscf.py [hunk 1]
```diff
@@ -41,6 +41,34 @@ def test_vqecasci_h2o_2e_2o():
...
+        2,
+        2,
+        optimizer=LBFGS(),
+        ansatz=Ansatz.GateFabric,
+        layers=2,
+        excitation_number=3,
+    )
+    mc.kernel()
+
+    # refmc = mcscf.CASCI(mf, 4, 4)
```
Comment (yyoshiyy) [inline tests/test_vqe_vqemcscf.py]: > The test was indeed not set up correctly. It only produced the ground state energy.

Thanks for noticing.
So was the corresponding CASCI setting of (4,4) also incorrect?
It is very minor, but can you fix them?
```
# refmc = mcscf.CASCI(mf, 4, 4)
# refmc.fcisolver.nroots = 5
```

**Review by yyoshiyy — CHANGES_REQUESTED** [review]
Thank you for fixing it.
I add one very minor comment, and it should be sufficient if this is fixed.

**Comment by yyoshiyy** [comment]
The following SSVQE calculation

```jsx
# Run CASCI with VQE with CAS(2e, 2o)
vqe_casci = VQECASCI(mf, 2, 2, optimizer=LBFGS(), ansatz=Ansatz.GateFabric, 
                     layers=2, excitation_number=5)
vqe_casci.kernel()
```

gives the strangely degenerated result:

```jsx
Calculated electronic state
state   Energy(Hartree) Excitation energy(eV)   Occupancy indicates
S0      -74.965695      0.0000                  
S1      -74.567104      10.8462                 1 -> 2
S2      -74.528864      11.8868                 0 -> 2
S3      -74.528864      11.8868                 1 -> 3
S4      -74.567104      10.8462                 0 -> 3
S5      -73.792297      31.9298                 0,1 -> 2,3
```

This is due to the how to generate initial states of SSVQE.

This part (in `vqeci.py`):

```jsx
        # Set initial Quantum State
        for m in range(self.n_electron, 2 * self.n_electron + 1):
            if comb(m, self.n_electron, exact=True) >= self.excitation_number + 1:
                break
        else:
            raise Exception("excitation_number is too large")

        occ_indices_lst = sorted(
            list(combinations(range(m), self.n_electron)),
            key=lambda lst: sum([2**a for a in lst]),
        )[: self.excitation_number + 1]
        self.occ_indices_lst = occ_indices_lst

        state_mapper = self.fermion_qubit_mapping.get_state_mapper(
            2 * self.n_orbitals, self.n_electron
        )
        self.initial_states = [state_mapper(x) for x in occ_indices_lst]
```

**Comment by yyoshiyy** [comment]
In the case of 

```jsx
n_electron=2, n_orbitals=2, excitation_number=5, 
fermion_qubit_mapping=jordan_wigner,
```

the initial states become

```jsx
Initial State 0: ComputationalBasisState(qubit_count=4, bits=0b0011, phase=0.0)
Initial State 1: ComputationalBasisState(qubit_count=4, bits=0b0101, phase=0.0)
Initial State 2: ComputationalBasisState(qubit_count=4, bits=0b0110, phase=0.0)
Initial State 3: ComputationalBasisState(qubit_count=4, bits=0b1001, phase=0.0)
Initial State 4: ComputationalBasisState(qubit_count=4, bits=0b1010, phase=0.0)
Initial State 5: ComputationalBasisState(qubit_count=4, bits=0b1100, phase=0.0)
```

The obtained initial states 2 and 3 are not spin-adapted.

These states should be 

$$
\frac{1}{\sqrt{2}}(|0110\rangle + |1001\rangle)
$$

$$
\frac{1}{\sqrt{2}}(|0110\rangle - |1001\rangle)
$$

**Comment by yyoshiyy** [comment]
- Writing a function to prepare spin-adapted initial states for general cases is cumbersome.
    - Therefore, I would like to allow users to manually select the initial state.
    - If nothing is provided, the current method will automatically generate the initial state.
- The following is an example of how to create an initial state that works well.

```jsx
from quri_parts.core.state import comp_basis_superposition
from math import pi

state_2 = self.initial_states[2]
state_3 = self.initial_states[3]

# Define spin-adapted linear combinations
state_2_spin_adapted = comp_basis_superposition(state_2, state_3, theta=+pi/4, phi=0.0)
state_3_spin_adapted = comp_basis_superposition(state_2, state_3, theta=-pi/4, phi=0.0)

# Replace initial states with spin-adapted states
self.initial_states[2] = state_2_spin_adapted
self.initial_states[3] = state_3_spin_adapted
```

The remaining states are the same as before.

**Comment by yyoshiyy** [comment]
As a summary, the specific revisions are as follows: 
1. Please refactor the original initial state generator into a function. 
2. Allow users to set the initial state as they wish. 
3. If nothing is set, have it call the initial state generator. 
4. For the test, please use the initial state I suggested.

**Review by yyoshiyy — APPROVED** [review]

**Comment by yyoshiyy** [comment]
We will address this debugging by creating a new issue.

**Pull request merged**